File: src/EtherGizmos.Shipyard.Models/Database/Package.cs
using EtherGizmos.Common.Abstractions;

namespace EtherGizmos.Shipyard.Database;

public class Package : Auditable, IEntity
{
    public virtual int Id { get; set; }

    public virtual int CarrierId { get; set; }

    public virtual Carrier Carrier { get; set; } = null!;

    public virtual string TrackingNumber { get; set; } = null!;

    public virtual string? Contents { get; set; }

    public virtual DateTimeOffset? EstimatedDeliveryAt { get; set; }

    public virtual DateTimeOffset LastPollAt { get; set; }

    public virtual DateTimeOffset NextPollAt { get; set; }

    public virtual int LastStatusTypeId { get; set; }

    public virtual StatusType LastStatusType { get; set; } = null!;

    public virtual bool IsDelivered { get; set; }

    public virtual List<TrackingUpdate> TrackingUpdates { get; set; } = [];

    public virtual Guid SecurableId { get; set; }

    public virtual Securable Securable { get; set; } = new() { Type = Enums.SecurableType.Package };
}

File: src/EtherGizmos.Shipyard.Web/src/app/features/package/models/package.ts
import { z } from "zod";
import { DateTimeZ } from "../../../shared/types/datetime/datetime";
import { AppValidators, formFactoryForModel } from "../../../shared/utilities/form/form.util";
import { CarrierZ } from "../../carrier/models/carrier";
import { StatusType } from "./status-type";
import { TrackingUpdateZ } from "./tracking-update";

export const PackageZ = z.object({
  id: z.number().int(),
  createdAt: DateTimeZ,
  modifiedAt: DateTimeZ,
  carrierId: z.number(),
  carrier: CarrierZ.nullish(),
  trackingNumber: z.string(),
  contents: z.string(),
  estimatedDeliveryAt: DateTimeZ.nullish(),
  lastPollAt: DateTimeZ,
  nextPollAt: DateTimeZ,
  lastStatusType: z.nativeEnum(StatusType),
  isDelivered: z.boolean(),
  trackingUpdates: z.array(TrackingUpdateZ).nullish(),
});

export interface Package extends z.infer<typeof PackageZ> { }

export type PackageF = Omit<Package, "carrier" | "trackingUpdates">;

export const packageForm = formFactoryForModel<PackageF>(($form, model) => ({
  id: [model.id],
  createdAt: [model.createdAt],
  modifiedAt: [model.modifiedAt],
  carrierId: [model.carrierId, AppValidators.required],
  trackingNumber: [model.trackingNumber, AppValidators.required],
  contents: [model.contents],
  estimatedDeliveryAt: [model.estimatedDeliveryAt],
  lastPollAt: [model.lastPollAt],
  nextPollAt: [model.nextPollAt],
  lastStatusType: [model.lastStatusType],
  isDelivered: [model.isDelivered],
}));
